### ISSUE TITLE
Extension traits should not be used as trait bounds

### ISSUE THREAD

**Issue by Darksonn** [rust-lang/rust-clippy#9390]
### What it does

It detects when an extension trait is used as a trait bound. The base trait should be used instead. For example:
```Rust
async fn takes_async_read<R: AsyncReadExt>(read: R) { ... }
```
In this case, it should suggest writing `R: AsyncRead` instead.

To define "extension trait", one possibility is any trait of the following form:
```Rust
trait MyTraitExt: MyTrait { .. }

impl<T: MyTrait + ?Sized> MyTraitExt for T {}
```

### Lint Name

extension-trait-as-trait-bound

### Category

suspicious

### Advantage

- It is strictly more powerful. The compiler will deduce that the type also implements the extension trait, and its impossible for something to implement the extension trait without implementing the base trait.
- The base trait is required in various cases. For example, you do not get the `impl AsyncRead for BufReader<R>` if you only have `R: AsyncReadExt`.
- It is conceptually incorrect.

### Drawbacks

Somewhat unclear how "extension trait" should be defined. For example, should `+ ?Sized` be required for it to be an extension trait? 

It's also unclear if there are cases where this would be a false positive.

### Example

```rust
async fn takes_async_read<R: AsyncReadExt>(read: R) { ... }
```

Could be written as:

```rust
async fn takes_async_read<R: AsyncRead>(read: R) { ... }
```

**Comment by lukaslueg** [comment]
It's not clear to me why a trait bound like this should be considered bad practice. `AsyncReadExt: AsyncRead`, and `takes_async_read` may require both.

**Comment by Darksonn** [comment]
Due to the blanket impl, `R: AsyncRead` implies `R: AsyncReadExt`, so they are equivalent mechanically. I suppose "bad style" would be a better category.

Here's another example:
```Rust
fn takes_iterator<I: itertools::Itertools>(iter: I) { .. }
```
However, people tend to do this more often with `AsyncRead`.

**Comment by lukaslueg** [comment]
Ah, now I see your point. As any `R` that is `R: AsyncReadExt` has to be `R: AsyncRead` (due to `pub trait AsyncReadExt: AsyncRead`), and as all `R: AsyncRead + ?Sized` are `AsyncReadExt` due to `impl<R: AsyncRead + ?Sized> AsyncReadExt for R`, a trait bound `R: AsyncReadExt` is never more powerful than a `R: AsyncRead`.

**Comment by Darksonn** [comment]
I've only ever seen people make the mistake with traits called `StreamExt`, `AsyncReadExt`, and `AsyncWriteExt`. (Though note that all of those names have multiple instances of them in different crates.)

**Comment by Nemo157** [comment]
I just realized there is actually a counter-example that probably fits the extension trait form: traits used as aliases to simplify bounds, e.g. [`DeserializeOwned`](https://docs.rs/serde/latest/serde/de/trait.DeserializeOwned.html).

Those are normally method-less marker traits, so if automatic detection of extension traits based on the form is done it should probably exclude any that lack methods.
